code/segmentedtickerplant/stplog.q

```diff
@@ -0,0 +1,238 @@
+ };
+
+// Custom mode - mixed periodic/tabular mode
+// Tables are defined as periodic, tabular, tabperiod or none in config file stpcustom.csv
+// Tables not specified in csv are not logged
+logname[`custom]:{[dir;tab;p]
+  logname[custommode tab][dir;tab;p]
+ };
+
+///////////////////////////////////////////////////////////////////////////////////////
+
+// Update and timer functions in three batch modes ////////////////////////////////////
+// preserve pre-existing definitions
+upd:@[value;`.stplg.upd;enlist[`]!enlist ()];
+zts:@[value;`.stplg.zts;enlist[`]!enlist ()];
+
+// Number of update messages received for each table
+msgcount:rowcount:tmpmsgcount:tmprowcount:enlist[`]!enlist ()
+
+// Sequence number
+seqnum:0
+
+// Functions to add columns on updates
+updtab:@[value;`.stplg.updtab;enlist[`]!enlist {(enlist(count first x)#y),x}]
+
+// If set to autobatch, publish and write to disk will be run in batches
+// insert to table in memory, on a timer flush the table to disk and publish, update counts
+upd[`autobatch]:{[t;x;now]
+  t insert updtab[t] . (x;now);
+ };
+
+zts[`autobatch]:{
+  {[t]
+    if[count value t;
+      `..loghandles[t] enlist (`upd;t;value flip value t);
+      @[`.stplg.msgcount;t;+;1];
+      @[`.stplg.rowcount;t;+;count value t];
+      .stpps.pubclear[t]];
+  }each .stpps.t;
+ };
+
+// Standard batch mode - write to disk immediately, publish in batches
+upd[`defaultbatch]:{[t;x;now]
+  t insert x:.stplg.updtab[t] . (x;now);
+  `..loghandles[t] enlist(`upd;t;x);
+  // track tmp counts, and add these after publish
+  @[`.stplg.tmpmsgcount;t;+;1];
+  @[`.stplg.tmprowcount;t;+;count first x];
+ };
+
+zts[`defaultbatch]:{
+  // publish and clear all tables, increment counts
+  .stpps.pubclear[.stpps.t];
+  // after data has been published, updated the counts
+  .stplg.msgcount+:.stplg.tmpmsgcount;
+  .stplg.rowcount+:.stplg.tmprowcount;
+  // reset temp counts
+  .stplg.tmpmsgcount:.stplg.tmprowcount:()!();
+ };
+
+// Immediate mode - publish and write immediately
+upd[`immediate]:{[t;x;now]
+  x:updtab[t] . (x;now);
+  `..loghandles[t] enlist(`upd;t;x);
+  @[`.stplg.msgcount;t;+;1];
+  @[`.stplg.rowcount;t;+;count first x];
+  .stpps.pub[t;x]
+ };
+
+zts[`immediate]:{}
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+// Functions to obtain logs for client replay ////////////////////////////////////////
+// replaylog called from client-side, returns nested list of logcounts and lognames
+replaylog:{[t]
+  getlogs[replayperiod][t]
+ }
+
+getlogs:enlist[`]!enlist ()
+
+// If replayperiod set to `period, only replay logs for current logging period
+getlogs[`period]:{[t]
+  distinct flip (.stplg.msgcount;exec tbl!logname from `..currlog where tbl in t)@\:t
+ };
+
+// If replayperiod set to `day, replay all of today's logs
+getlogs[`day]:{[t]
+  lnames:distinct uj/[{
+    select seq,tbls,logname,msgcount from .stpm.metatable where x in/: tbls
```

Comment:
make it `msgcount:0W` to ensure that any previous log files are always replayed in full